precompiles/batch/src/tests.rs

```diff
@@ -0,0 +1,922 @@
+		batch_out_of_gas(&precompiles(), Action::BatchSomeUntilFailure)
+			.expect_log(log_subcall_failed(Precompile, 0))
+			.execute_returns(Vec::new())
+	})
+}
+
+#[test]
+fn batch_all_out_of_gas() {
+	ExtBuilder::default().build().execute_with(|| {
+		batch_out_of_gas(&precompiles(), Action::BatchAll).execute_error(ExitError::OutOfGas)
+	})
+}
+
+fn batch_incomplete(
+	precompiles: &TestPrecompiles<Runtime>,
+	action: Action,
+) -> PrecompilesTester<TestPrecompiles<Runtime>> {
+	let mut counter = 0;
+
+	let (return_log_cost, total_call_cost) = costs();
+
+	precompiles
+		.prepare_test(
+			Alice,
+			Precompile,
+			EvmDataWriter::new_with_selector(action)
+				.write(vec![
+					Address(Bob.into()),
+					Address(Charlie.into()),
+					Address(Alice.into()),
+				])
+				.write(vec![U256::from(1u8), U256::from(2u8), U256::from(3u8)])
+				.write(vec![Bytes::from(b"one".as_slice())])
+				.write::<Vec<U256>>(vec![])
+				.build(),
+		)
+		.with_target_gas(Some(300_000))
+		.with_subcall_handle(move |subcall| {
+			let Subcall {
+				address,
+				transfer,
+				input,
+				target_gas,
+				is_static,
+				context,
+			} = subcall;
+
+			// Called from the precompile caller.
+			assert_eq!(context.caller, Alice.into());
+			assert_eq!(is_static, false);
+
+			match address {
+				a if a == Bob.into() => {
+					assert_eq!(counter, 0, "this is the first call");
+					counter += 1;
+
+					assert_eq!(
+						target_gas,
+						Some(300_000 - total_call_cost),
+						"batch forward all gas"
+					);
+					let transfer = transfer.expect("there is a transfer");
+					assert_eq!(transfer.source, Alice.into());
+					assert_eq!(transfer.target, Bob.into());
+					assert_eq!(transfer.value, 1u8.into());
+
+					assert_eq!(context.address, Bob.into());
+					assert_eq!(context.apparent_value, 1u8.into());
+
+					assert_eq!(&input, b"one");
+
+					SubcallOutput {
+						reason: ExitReason::Succeed(ExitSucceed::Returned),
+						output: Vec::new(),
+						cost: 13,
+						logs: vec![
+							LogsBuilder::new(Bob.into()).log1(H256::repeat_byte(0x11), vec![])
+						],
+					}
+				}
+				a if a == Charlie.into() => {
+					assert_eq!(counter, 1, "this is the second call");
+					counter += 1;
+
+					assert_eq!(
+						target_gas,
+						Some(300_000 - 13 - total_call_cost * 2),
+						"batch forward all gas"
+					);
+					let transfer = transfer.expect("there is a transfer");
+					assert_eq!(transfer.source, Alice.into());
+					assert_eq!(transfer.target, Charlie.into());
+					assert_eq!(transfer.value, 2u8.into());
+
+					assert_eq!(context.address, Charlie.into());
+					assert_eq!(context.apparent_value, 2u8.into());
+
+					assert_eq!(&input, b"");
+
+					SubcallOutput {
+						reason: ExitReason::Revert(ExitRevert::Reverted),
+						output: b"Revert message".to_vec(),
+						cost: 17,
+						logs: vec![],
+					}
+				}
+				a if a == Alice.into() => {
+					assert_eq!(counter, 2, "this is the third call");
+					counter += 1;
```

Comment:
once here not necessary to inc the counter

Reply:
Ensure there is not an additional call to this same address for some reason.

Reply:
Fair enough